File: q4/haskell/PingPing.hs
import Control.Concurrent.MVar
import Control.Concurrent
import Control.Monad
import System.IO

-- Agora o waitThread irá esperar uma única thread terminar.
waitThread :: MVar () -> IO ()
waitThread = takeMVar

sender :: MVar String -> Int -> IO ()
sender queue n =
    forM_ [1..n] $ \i -> do
        putMVar queue "Ping"
        putStrLn "Sent: Ping"
        threadDelay 2

receiver :: MVar String -> Int -> IO ()
receiver queue n =
    forM_ [1..n] $ \i -> do
        msg <- takeMVar queue
        putStrLn ("Received: " ++ msg)
        threadDelay 2

main :: IO ()
main = do
    queue <- newEmptyMVar

    putStrLn "Digite valor de N:"
    n <- readLn

    -- Criamos as barreiras para sinalizar quando as threads terminam
    senderDone <- newEmptyMVar
    receiverDone <- newEmptyMVar

    forkIO (sender queue n >> putMVar senderDone ())    -- Quando a thread 'sender' terminar, ela sinaliza
    forkIO (receiver queue n >> putMVar receiverDone ()) -- Quando a thread 'receiver' terminar, ela sinaliza

    -- Espera ambas as threads terminarem
    waitThread senderDone
    waitThread receiverDone

File: q4/java/PingPing.java
package java;
import java.util.concurrent.*;
import java.util.Scanner;

public class PingPing {
    private static BlockingQueue<String> queue = new LinkedBlockingQueue<>();

    public static void main(String[] args) throws InterruptedException {
        Scanner scanner = new Scanner(System.in);

        System.out.println("Digite valor de N:");
        int N = scanner.nextInt();
        scanner.close();

        Thread producer = new Thread(() -> {
            try {
                for (int i = 0; i < N; i++) {
                    queue.put("Ping");
                    System.out.println("Sent: Ping");
                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        });

        Thread consumer = new Thread(() -> {
            try {
                for (int i = 0; i < N; i++) {
                    String msg = queue.take();
                    System.out.println("Received: " + msg);
                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        });

        consumer.start();
        producer.start();

        consumer.join();
        producer.join();
    }
}
